File: frontend/spatial_canvas_app/lib/models/spatial_object.dart
class SpatialObject {
  final String id;
  double x;
  double y;
  final String color;
  final String shape;
  final double size;

  SpatialObject({
    required this.id,
    required this.x,
    required this.y,
    required this.color,
    required this.shape,
    required this.size,
  });

  factory SpatialObject.fromJson(Map<String, dynamic> json) {
    final loc = json['loc'] as List<dynamic>;
    return SpatialObject(
      id: json['_id'] as String,
      x: (loc[0] as num).toDouble(),
      y: (loc[1] as num).toDouble(),
      color: json['color'] as String? ?? '#3498db',
      shape: json['shape'] as String? ?? 'circle',
      size: (json['size'] as num?)?.toDouble() ?? 8.0,
    );
  }
}

File: backend/src/models/SpatialObject.js
const mongoose = require('mongoose');

const SpatialObjectSchema = new mongoose.Schema({
  loc: {
    type: [Number], // [x, y] — required shape for a 2d index
    required: true,
    validate: {
      validator: (v) => Array.isArray(v) && v.length === 2,
      message: 'loc must be an array of exactly [x, y]',
    },
  },
  color: {
    type: String,
    default: '#3498db',
  },
  shape: {
    type: String,
    enum: ['circle', 'square', 'triangle'],
    default: 'circle',
  },
  size: {
    type: Number,
    default: 8,
  },
}, { timestamps: true }); 

SpatialObjectSchema.index({ loc: '2d' }, { min: -10000, max: 10000 });

module.exports = mongoose.model('SpatialObject', SpatialObjectSchema);